File: Sigmath/Sigmath/parser.cs
using System;
using System.Data;
using System.Linq.Expressions;

namespace Sigmath
{
    public class ExpressionParser
    {
        private readonly DataTable dataTable;

        public ExpressionParser(DataTable dataTable)
        {
            this.dataTable = new DataTable();
        }

        public Expression Parse(string expression)
        {
            try
            {
                return Expression.Constant(new DataTable().Compute(expression, null), typeof(double));
            }
            catch
            {
                throw new ArgumentException("Invalid expression");
            }
        }

        public Func<double> Compile(Expression expression)
        {
            try
            {
                var lambda = Expression.Lambda<Func<double>>(expression);
                var compiled = lambda.Compile();
                return compiled;
            }
            catch
            {
                throw new ArgumentException("Invalid expression");
            }
        }
    }
}

File: Sigmath.ConvertedToRuby/parser.rb
require "mscorlib"

require "mscorlib"
require "System, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"
require "System.Data, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"
require "System.Linq.Expressions, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a"

module Sigmath
	class ExpressionParser
		def initialize(dataTable)
			self.@dataTable = dataTable
		end

		def Parse(expression)
			begin
				return Expression.Constant(@dataTable.Compute(expression, nil), System::Double.to_clr_type)
			rescue  => 
				raise ArgumentException.new("Invalid expression")
			ensure
			end
		end

		def Compile(expression)
			begin
				lambda = Expression.Lambda(expression)
				compiled = lambda.Compile()
				return compiled
			rescue  => 
				raise ArgumentException.new("Invalid expression")
			ensure
			end
		end
	end
end